src/variant.h

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) 2017-2019 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_VARIANT_H
+#define BITCOIN_VARIANT_H
+
+#include <utility>
+
+#include <boost/variant.hpp>
+
+//! Substitute for C++17 std::variant
+template <class... Types>
+class Variant
+{
+private:
+    boost::variant<Types...> m_var;
+
+public:
+    Variant()
+    {
+        m_var = {};
+    }
+    Variant(const Variant& other)
+    {
+        m_var = other.m_var;
+    }
+    Variant(Variant& other)
+    {
+        m_var = other.m_var;
+    }
+    Variant(Variant&& other)
+    {
+        m_var = std::move(other.m_var);
+    }
+    template <class T>
+    Variant(const T& value)
+    {
+        m_var = value;
+    }
+    template <class T>
+    Variant(T&& value)
+    {
+        m_var = std::move(value);
+    }
+
+    int which() const
+    {
+        return m_var.which();
+    }
+
+    template <class T>
+    T* get_if()
+    {
+        return boost::get<T>(&m_var);
+    }
+    template <class T>
+    const T* get_if() const
+    {
+        return boost::get<T>(&m_var);
+    }
+    template <class T>
+    T& get()
+    {
+        return boost::get<T>(m_var);
+    }
+    template <class T>
+    const T& get() const
+    {
+        return boost::get<T>(m_var);
+    }
+
+    Variant& operator=(const Variant& other)
+    {
+        m_var = other.m_var;
+        return *this;
+    }
+    Variant& operator=(Variant&& other)
```

Comment:
In commit "Add a new Variant type" (85882c01d20add49f7b1f8b741e078968ec8080b)

Maybe add a template `operator=`, too. https://en.cppreference.com/w/cpp/utility/variant/operator%3D shows this as the third overload for `operator=`

Reply:
seems that also here(like the constructor) boost adds both rlvaue and lvalue operators. i'll add just the rvalue like std::variant does. (although I don't quite understand why only that)

Reply:
> seems that also here(like the constructor) boost adds both rlvaue and lvalue operators. i'll add just the rvalue like std::variant does. (although I don't quite understand why only that)

The reason why only a `T&&` overload is needed is that when `T` is a template type, there are template deduction rules that let `T` be deduced to be an lvalue reference, so the function can be called with both lvalue and rvalue arguments. (This is needed for "perfect forwarding"). If `T` was a normal type and not a template type, you would need to have both `T&` and `T&&` overloads to accept lvalues and rvalues.

Reply:
argh This also introduces the same problem as in https://github.com/bitcoin/bitcoin/pull/17953#discussion_r370356157

Reply:
Done